check_upgrade/check_upgrade.go

```diff
@@ -0,0 +1,232 @@
+		for j, node2 := range aclNodes {
+			if i != j && node1.DgraphXID == node2.DgraphXID {
+				duplicates = append(duplicates, node1)
+				break
+			}
+		}
+	}
+
+	return duplicates
+}
+
+func getDuplicateNodes(ctx context.Context, dg *dgo.Dgraph) ([]ACLNode, error) {
+	query := `{ 
+		nodes(func: has(dgraph.xid)) {
+			       uid
+			       dgraph.xid
+                   dgraph.type
+		        }
+	}`
+
+	resp, err := dg.NewTxn().Query(ctx, query)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while querying dgraph for duplicate nodes")
+	}
+
+	type Nodes struct {
+		Nodes []ACLNode `json:"nodes"`
+	}
+	var result Nodes
+	if err := json.Unmarshal([]byte(resp.Json), &result); err != nil {
+		return nil, errors.Wrapf(err, "while unmarshalling response: %v", string(resp.Json))
+
+	}
+
+	return findDuplicateNodes(result.Nodes), nil
+}
+
+func printDuplicates(entityType string, ns uint64, nodes []ACLNode) {
+	if len(nodes) > 0 {
+		fmt.Printf("Found duplicate %ss in namespace: #%v\n", entityType, ns)
+		for _, node := range nodes {
+			fmt.Printf("dgraph.xid %v , Uid: %v\n", node.DgraphXID, node.UID)
+		}
+	}
+}
+
+func init() {
+	CheckUpgrade.Cmd = &cobra.Command{
+		Use:   "checkupgrade",
+		Short: "Run the checkupgrade tool",
+		Long:  "The checkupgrade tool is used to check for duplicate dgraph.xid's in the Dgraph database before upgrade.",
+		Run: func(cmd *cobra.Command, args []string) {
+			run()
+		},
+		Annotations: map[string]string{"group": "tool"},
+	}
+	CheckUpgrade.Cmd.SetHelpTemplate(x.NonRootTemplate)
+	flag := CheckUpgrade.Cmd.Flags()
+	flag.String(alphaGrpc, "127.0.0.1:9080",
+		"Dgraph Alpha gRPC server address")
+
+	flag.String(alphaHttp, "http://127.0.0.1:8080", "Draph Alpha HTTP(S) endpoint.")
+}
+
+func printDupDeleteMutation(nodes []ACLNode) {
+	if len(nodes) == 0 {
+		return
+	}
+	var deleteMutation strings.Builder
+	deleteMutation.WriteString("{\n")
+	deleteMutation.WriteString(" delete {\n")
+	for _, node := range nodes {
+		deleteMutation.WriteString(fmt.Sprintf("<%v> <%v> \"%v\" .\n", node.UID, "dgraph.xid", node.DgraphXID))
+	}
+	deleteMutation.WriteString(" }\n")
+	deleteMutation.WriteString(" }\n")
+
+	fmt.Printf("\ndelete duplicate nodes using following mutation : \n%v", deleteMutation.String())
+}
+
+func run() {
+	if err := checkUpgrade(); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+	}
+}
+
+func checkUpgrade() error {
+	fmt.Println("Running check-upgrade tool")
+
+	cmdInput := parseInput()
+
+	gc, hc, err := setupClients(cmdInput.alphaGrpc, cmdInput.alphaHttp)
+	if err != nil {
+		return errors.Wrapf(err, "while setting up clients")
+	}
+
+	hc.LoginIntoNamespace(dgraphtest.DefaultUser, dgraphtest.DefaultPassword, x.GalaxyNamespace)
+	if err != nil {
+		return errors.Wrapf(err, "while logging into namespace: %v", x.GalaxyNamespace)
+	}
+
+	namespaces, err := hc.ListNamespaces()
+	if err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+	for _, ns := range namespaces {
```

Comment:
Take these as args as well

Reply:
from user?

Reply:
yes, we should take the username and password as an argument from the user. I am thinking if we have a a galaxy user, that should allow us to read all the namespaces.